### PR TITLE
Add Weapon Offset Customization

### PR DESCRIPTION
**Pull request by Berthalamew**

### CODE REVIEW

File: xlive/H2MOD/Modules/Shell/Config.cpp [hunk 1]
```diff
@@ -48,8 +48,8 @@ bool H2Config_mouse_uniform = false;
 float H2Config_controller_sens = 0;
 bool H2Config_controller_modern = false;
 H2Config_Deadzone_Type H2Config_Controller_Deadzone = H2Config_Deadzone_Type::Axial;
-float H2Config_Deadzone_A_X = 1.0f;
-float H2Config_Deadzone_A_Y = 1.0f;
+float H2Config_Deadzone_A_X = 26.0f;
```
Comment (nukeulater) [inline xlive/H2MOD/Modules/Shell/Config.cpp:51]: any reason for modifying this?

File: xlive/H2MOD/Modules/WeaponOffsets/WeaponOffsetConfig/WeaponOffsetConfig.cpp [hunk 2]
```diff
@@ -0,0 +1,61 @@
+#include "stdafx.h"
+#include "WeaponOffsetConfig.h"
+
+char* path = strcat
```
Comment (nukeulater) [inline xlive/H2MOD/Modules/WeaponOffsets/WeaponOffsetConfig/WeaponOffsetConfig.cpp]: this should preferably be executed in the scope of the function requiring the file path, not globally

File: xlive/H2MOD/Modules/WeaponOffsets/WeaponOffsetConfig/WeaponOffsetConfig.cpp [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+void SaveWeaponOffsetConfig(const s_weapon_custom_offset customOffsets[], bool defaultOffsets)
+{
+    FILE* file = NULL;
+
+    file = fopen(path, "w");
+    if (file != NULL)
+    {
+        if (defaultOffsets == false)
+        {
+            for (byte i = BattleRifle; i != Sniper + 1; i++)
```
Comment (nukeulater) [inline xlive/H2MOD/Modules/WeaponOffsets/WeaponOffsetConfig/WeaponOffsetConfig.cpp]: weird and unusual way of looping through some array, please fix

File: xlive/H2MOD/Modules/WeaponOffsets/WeaponOffsetConfig/WeaponOffsetConfig.h [hunk 4]
```diff
@@ -0,0 +1,34 @@
+#include "Blam/Math/real_math.h"
+#include "Blam\Cache\TagGroups\weapon_definition.hpp"
+
+struct s_weapon_custom_offset {
+	const real_vector3d DefaultOffset;
```
Comment (nukeulater) [inline xlive/H2MOD/Modules/WeaponOffsets/WeaponOffsetConfig/WeaponOffsetConfig.h]: variable names should start with lower case characters, upper reserved for class/struct member functions

File: xlive/H2MOD/Modules/WeaponOffsets/WeaponOffsetConfig/WeaponOffsetConfig.h [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+	const char WeaponPath[128];
+	real_vector3d ModifiedOffset;
+	s_weapon_group_definition* tag;
+};
+
+enum Weapons
+{
+	BattleRifle,
```
Comment (nukeulater) [inline xlive/H2MOD/Modules/WeaponOffsets/WeaponOffsetConfig/WeaponOffsetConfig.h]: quite ambiguous enum elements here (e.g BattleRifle could be used in other parts of the code, not just within weapon offset modification context), and enum name should use `e_` prefix with lowercase name after

File: xlive/H2MOD/Modules/WeaponOffsets/WeaponOffsetConfig/WeaponOffsetConfig.h [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+	PlasmaRifle,
+	RocketLauncher,
+	SentinelBeam,
+	Shotgun,
+	SMG,
+	Sniper
+};
+
+void ReadWeaponOffsetConfig(s_weapon_custom_offset* WeaponOffsets);
+void SaveWeaponOffsetConfig(const s_weapon_custom_offset customOffsets[], bool defaultOffsets);
```
Comment (nukeulater) [inline xlive/H2MOD/Modules/WeaponOffsets/WeaponOffsetConfig/WeaponOffsetConfig.h]: prefer using pointers over arrays for function arguments, array in this case does decay to pointers but we should preferably be consistent across the codebase

File: xlive/H2MOD/Modules/WeaponOffsets/WeaponOffsets.cpp [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+		void ApplyOffset(Weapons weapon)
+		{
+			if (customOffsets[weapon].tag != nullptr)
+			{
+				customOffsets[weapon].tag->first_person_weapon_offset = customOffsets[weapon].ModifiedOffset;
+			}
+		}
+		void MapLoad()
+		{
+			for (byte i = BattleRifle; i != Sniper + 1; i++)
```
Comment (nukeulater) [inline xlive/H2MOD/Modules/WeaponOffsets/WeaponOffsets.cpp]: same thing here, weird and unusual way of looping

File: xlive/H2MOD/Modules/WeaponOffsets/WeaponOffsetConfig/WeaponOffsetConfig.cpp (hunk 2 above)
Comment (nukeulater) [reply xlive/H2MOD/Modules/WeaponOffsets/WeaponOffsetConfig/WeaponOffsetConfig.cpp]: also could probably use `std::filesystem` library here

File: xlive/H2MOD/Modules/WeaponOffsets/WeaponOffsetConfig/WeaponOffsetConfig.cpp [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+);
+
+void ReadWeaponOffsetConfig(s_weapon_custom_offset *WeaponOffsets)
+{
+    FILE* file = NULL;
+
+    file = fopen(path, "r");
+    if (file != NULL)
+    {
+        for (byte i = BattleRifle; i != Sniper + 1; i++)
```
Comment (nukeulater) [inline xlive/H2MOD/Modules/WeaponOffsets/WeaponOffsetConfig/WeaponOffsetConfig.cpp]: weird and unusual way of looping through some array, please fix

File: xlive/H2MOD/Modules/Shell/Config.cpp (hunk 1 above)
Comment (Berthalamew) [reply xlive/H2MOD/Modules/Shell/Config.cpp:51]: The default deadzone ingame is 26, for some reason it's set to 1 in the config so I changed it to match that

File: xlive/H2MOD/Modules/WeaponOffsets/WeaponOffsets.cpp [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+				OffsetMenu(static_cast<Weapons>(selectedOption), "##OffsetX", weapon_offset_x,
+					customOffsets[selectedOption].ModifiedOffset.i, customOffsets[selectedOption].DefaultOffset.i);
+				OffsetMenu(static_cast<Weapons>(selectedOption), "##OffsetY", weapon_offset_y,
+					customOffsets[selectedOption].ModifiedOffset.j, customOffsets[selectedOption].DefaultOffset.j);
+				OffsetMenu(static_cast<Weapons>(selectedOption), "##OffsetZ", weapon_offset_z,
+					customOffsets[selectedOption].ModifiedOffset.k, customOffsets[selectedOption].DefaultOffset.k);
+			}
+		}
+
+		char* GetString(e_weapon_offsets_string string, const std::string& id)
```
Comment (nukeulater) [inline xlive/H2MOD/Modules/WeaponOffsets/WeaponOffsets.cpp]: if this function returns a string buffer that shouldn't be modified, it should return `const char*`. I noticed you just copied the code from another place expecting it to be fine. you might want to fix this there as well and remove the casting that removes the constness of the string buffers.

**Pull request merged**